builder/dockerfile/evaluator.go

```diff
@@ -194,12 +236,14 @@ type dispatchState struct {
 	imageID    string
 	baseImage  builder.Image
 	stageName  string
+	shlex      *ShellLex
+	builder    *Builder
+	source     builder.Source
```

Comment:
I don't know if it makes sense to have all the dispatch logic on a single struct. They are not really behaviours of an object. Having a request object that is passed to each is basically the same idea, but I think it makes the relationship between the dispatch functions more obvious.